guides/ci/travis.md

```diff
@@ -0,0 +1,252 @@
+
+## Installing binary dependencies
+
+> It is important to mention that there are many ways to achieve this and **it will heavily depend on our development workflow**.
+
+Our application or maybe some shards may required libraries and packages. In that case we will need to install them using `apt` (or similar).
+
+Here is a first example installing SQLite3 using Travis CI configuration file:
+
+```yaml
+# .travis.yml
+language: crystal
+crystal:
+  - latest
+
+before_install:
+  - sudo apt-get -y install libsqlite3-dev
+
+addons:
+  apt:
+    update: true
+
+script:
+  - crystal spec
+```
+
+#### Using Docker
+
+The same can be accomplished using a containerization service like Docker. Here is an example of a `Dockerfile` installing SQLite3:
+
+```dockerfile
+# Dockerfile
+FROM crystallang/crystal:latest
+
+RUN apt-get update && apt-get install -y libsqlite3-dev
+```
+
+And here is the Travis CI configuration file:
+
+```yml
+# .travis.yml
+language: bash
+
+services:
+  - docker
+
+before_install:
+  # build image using Dockerfile:
+  - docker build -t testing .
+
+script:
+  # run specs in the container
+  - docker run -v $PWD:/src -w /src testing crystal spec
+```
+
+**Note:** Dockerfile arguments can be used to use the same Dockerfile for latest, nightly or a specific version.
+
+## Using a database (example using MySQL)
+
+As you may notice, in the last configuration file we are using `docker` as a `service`. Travis CI may start different services as needed. For example, we may need [MySQL](https://docs.travis-ci.com/user/database-setup/#mysql).
+
+```yaml
+# .travis.yml
+language: crystal
+crystal:
+  - latest
+
+services:
+  - mysql
+
+script:
+  - crystal spec
+```
+
+Here is the new test file for testing against the database:
+
+```crystal
+# spec/simple_db_spec.cr
+require "./spec_helper"
+require "mysql"
+
+it "connects to the database" do
+  DB.connect ENV["DATABASE_URL"] do |cnn|
+    cnn.query_one("SELECT 'foo'", as: String).should eq "foo"
+  end
+end
+```
+
+When pushing this changes Travis CI will report the following error: `Unknown database 'test' (Exception)`, showing that we need to configure the MySQL service **and also setup the database**:
+
+> It's really **important** to notice that the lines we are adding to `.travis.yml` will depend exclusively on the development workflow we are using!
+> And remember that this is only an example using MySQL.
+
+```yaml
+# .travis.yml
+language: crystal
+crystal:
+  - latest
+
+env:
+  - DATABASE_URL="mysql://root@localhost/test"
+
+services:
+  - mysql
+
+before_install:
+  - mysql -u root --password="" < test-data/setup.sql
+
+script:
+  - crystal spec
+```
+
+We are [using a `setup.sql` script](https://andidittrich.de/2017/06/travisci-setup-mysql-tablesdata-before-running-tests.html) to create a more readable `.travis.yml`. The file `./test-data/setup.sql` looks like this:
+
+```sql
+-- setup.sql
+CREATE DATABASE IF NOT EXISTS test;
```

Comment:
Hardcoding the database name in the schema file is bad practice. It unnecessarily restricts the usage. For example, you can't have a production and test database on the same server.
Database creation should be handled with service setup so that shard finds the database at `DATABASE_URL` already existing.